src/tactic/induction.lean

```diff
@@ -495,10 +496,61 @@ meta def constructor_intros (generate_induction_hyps : bool)
 meta def ih_name (arg_name : name) : name :=
 mk_simple_name ("ih_" ++ arg_name.to_string)
 
-private meta def get_with_name : option name → option name
-| (some `_) := none
-| (some n) := some n
-| none := none
+/--
+Representation of a pattern in the `with n ...` syntax supported by
+`induction'` and `cases'`. We support the following patterns:
+
+- `auto` (`with _` or no `with` clause): use the name generated by the tactic.
+- `clear` (`with -`): clear this hypothesis and any hypotheses depending on it.
+- `exact n` (`with n`): use the name `n` for this hypothesis.
+-/
+@[derive has_reflect]
+meta inductive with_pattern
+| auto
+| clear
+| exact (n : name)
+
+namespace with_pattern
+
+open lean
+open lean.parser
+
+/-- Parser for a `with_pattern`. -/
+meta def parser : lean.parser with_pattern :=
+(tk "-" *> pure with_pattern.clear) <|>
+(tk "_" *> pure with_pattern.auto) <|>
+(with_pattern.exact <$> ident)
+
+/-- Parser for a `with` clause. -/
+meta def clause_parser : lean.parser (list with_pattern) :=
+(tk "with" *> many parser) <|> pure []
+
+/--
+`to_name_spec auto_candidates p` returns a description of how the hypothesis to
+which the `with_pattern` `p` applies should be named. If this function returns
+`none`, the hypothesis should be cleared. If it returns `some (inl n)`, it
+should receive exactly the name `n`, even if this shadows other hypotheses. If
+it returns `some (inr ns)`, it should receive the first unused name from `ns`.
+
+If `p = auto`, the `auto_candidates` tactic is run to determine candidate names
+for the hypothesis (from which the first fresh one is later chosen).
+`auto_candidates` must return a nonempty list.
+-/
+meta def to_name_spec (auto_candidates : tactic (list name)) :
+  with_pattern → tactic (option (name ⊕ list name))
+| auto := (some ∘ sum.inr) <$> auto_candidates
+| clear := pure none
+| (exact n) := pure $ some $ sum.inl n
+
+end with_pattern
+
+/--
+If `h` refers to a hypothesis, `clear_dependent_if_exists h` clears `h` clears
+`h` and any hypotheses which depend on it. Otherwise, the tactic does nothing.
+-/
+meta def clear_dependent_if_exists (h : name) : tactic unit := do
```

Comment:
Do you know how `rcases` does the clear? Are there reasons the code can't be shared?

Reply:
I'd love to reuse the `rcases` code, but it's quite entangled with `cases`. In particular, `rcases` relies on `cases` for the auto-generated names, so doesn't have any name generation logic of its own, while that's a major part of my implementation. It would certainly be possible to disentangle and generalise the `rcases` stuff, but I don't think the effort is justified. Lean 4 will require major changes to the design anyway and `induction'` is hardly used.